trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopFactory.java

```diff
@@ -0,0 +1,350 @@
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.concurrency.PartialOrderCegarLoop;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TAPreferences;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.FloydHoareAutomataReuse;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.InterpolantAutomaton;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.LanguageOperation;
+import de.uni_freiburg.informatik.ultimate.witnessparser.graph.WitnessEdge;
+import de.uni_freiburg.informatik.ultimate.witnessparser.graph.WitnessNode;
+
+/**
+ * A utility class that allows creating CEGAR loops for different programs (based on some common settings).
+ *
+ * @author Matthias Heizmann (heizmann@informatik.uni-freiburg.de)
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ *
+ * @param <L>
+ *            The type of transitions in programs analyzed by the created CEGAR loops
+ */
+public class CegarLoopFactory<L extends IIcfgTransition<?>> {
+
+	private static final boolean FORCE_FINITE_AUTOMATA_FOR_SEQUENTIAL_PROGRAMS = true;
+
+	private final Class<L> mTransitionClazz;
+	private final TAPreferences mPrefs;
+	private final Supplier<IPLBECompositionFactory<L>> mCreateCompositionFactory;
+	private final boolean mComputeHoareAnnotation;
+
+	private CegarLoopStatisticsGenerator mCegarLoopBenchmark;
+
+	public CegarLoopFactory(final Class<L> transitionClazz, final TAPreferences taPrefs,
+			final Supplier<IPLBECompositionFactory<L>> createCompositionFactory, final boolean computeHoareAnnotation) {
+		mTransitionClazz = transitionClazz;
+		mPrefs = taPrefs;
+		mCreateCompositionFactory = createCompositionFactory;
+		mComputeHoareAnnotation = computeHoareAnnotation;
+	}
+
+	/**
+	 * Creates a new CEGAR loop.
+	 *
+	 * @param services
+	 *            Ultimate services to use. In particular, this may be used to set a timeout.
+	 * @param name
+	 *            An identifier for the CEGAR loop
+	 * @param root
+	 *            The control flow graph of the analyzed program
+	 * @param errorLocs
+	 *            The error locations whose unreachability shall be proven
+	 * @param witnessAutomaton
+	 *            An (optional) witness automaton
+	 * @param rawFloydHoareAutomataFromFile
+	 *            A list of automata to use if a CEGAR loop with Floyd/Hoare automata reuse is created
+	 *
+	 * @return the newly created CEGAR loop
+	 */
+	public BasicCegarLoop<L, ?> constructCegarLoop(final IUltimateServiceProvider services, final DebugIdentifier name,
```

Comment:
There seem to be many incompatible settings hidden here. Perhaps a TODO or something would be nice. 

E.g., IncrementalInclusion does not work for concurrent programs, right?

Reply:
What would the TODO be?

> IncrementalInclusion does not work for concurrent programs, right?

I doubt it's been tested, but it might -- it will use the automata-based approach though.

In general I agree there is a problem that we don't seem to have any code that checks settings for incompatibility. Sometimes an error might be thrown later, or one setting quietly overrides the value of another (e.g. here, incremental inclusion implies that finite automata are used, even if we choose Petri nets). We should probably discuss a solution, but not in this PR. :+1: